### ISSUE TITLE
GCM works intermittently with Azure DevOps Server

### ISSUE THREAD

**Issue by mkolakow** [git-ecosystem/git-credential-manager#1440]
### Version

2.3.2

### Operating system

Windows

### OS version or distribution

Windows 10 Enterprise

### Git hosting provider(s)

Azure DevOps Server (TFS/on-prem)

### (Azure DevOps only) What format is your remote URL?

None

### Can you access the remote repository directly in the browser?

Yes, I can access the repository

### Expected behavior

I recently started this position and was told that Git used to work and now it doesn't for anyone (myself included!). So, I'm trying to help fix it and am running out of ideas. Here are the details of our environment: 

* We have a chip / token that must be inserted in order to log on to any server
* We then enter a PIN to log on (e.g. to my laptop)
* We then have to log on through a VPN (e.g. using Ivanti)
  * At this point, I am able to start Chrome and get to our ADO server
  * So, I assume this is all using Windows Auth
* We are using Azure DevOps Server on Prem Version Dev17.M153.5. 
* We are using Visual Studio 2022 for an IDE
* We are required to restart our computers every day
* If we want to use Git, we have to do the following:
  * Try a git command once, it fails
  * Try the same command again, and GCM pops up with a username/password window
  * Go to a company internal site and get issued a temporary 1-day password
  * Enter windows username and that one-day password into GCM
  * Git will work for the rest of the day
  * This "workflow" is valid for either VS 2022 or using git on command line

Obviously, we'd like to not have to do this. Once we are logged into our work computers, the expectation is that git would also work. 

My first approach was to try the following: 
1. I installed Git For Windows 2.42.0.2 to get the latest
2. I use the setting `usehttppath = true` in my git config

This worked! For about 4 days, then it just stopped. Also, it has not worked for anyone else on the team. Oddly enough, I somehow got it to work again for a few days. 

Please note: our ADO server git repo url is the following: 

> http://abc-teamfound:8080/tfs/ABC-IT%20Development/ABC%20ABC%20Web%20Apps/_git/ABC

In other words, I did note that they have it set up using http vs https. That's what clued me into try the usehttppath setting. 

I have also tried removing credentials from the Windows Credential Manager in order force GCM to query or refresh. I also tried setting `preserve = true` hoping that would somehow save credentials. 

Here are the settings from my git config
```
[credential "http://abc-teamfound:8080"]
	provider = generic
	usehttppath = true
[credential]
	trace = C:/Users/123456/git-credential.log
	helper = manager
```

I'm not sure what else to check or where else to go from here. Can someone assist? 

### Actual behavior

Here's an example of what I see 

```
$ git pull
fatal: Authentication failed for 'http://abc-teamfound:8080/tfs/ABC-IT%20Development/ABC%20ABC%20Web%20Apps/_git/ABC/'
```

### Logs

Attached is a log file where it failed to authenticate. I had already retrieved my one-day password, but it didn't work because with the `usehttppath = true` setting, it wouldn't prompt me for credentials. So, I set it to false, so I could at least get authenticated, and then set it back to true to see what would happen. Please note: it had been set to true and worked for all of Wed/Thur/Fri last week, including daily restarts of my laptop.

[logsAfterOneDayPasswordUseHttpPathTrue.txt](https://github.com/git-ecosystem/git-credential-manager/files/12918949/logsAfterOneDayPasswordUseHttpPathTrue.txt)

**Comment by ldennington** [comment]
Hi! Thanks for all the information and detailed logs. It looks like your environment is very locked down; our best guess so far is that there is a problem with Windows authentication between Git (`libcurl`) and TFS. GCM shows in the logs that Windows authentication is advertised by the server and instructs Git to attempt Kerberos/NTLM. This is failing and the fallback is the basic username and password prompt that you see.

GCM does not actually perform Windows authentication - it delegates this to Git/CURL. We may be able to help further, however, if we see the CURL logs from Git. You can obtain this by setting the following environment variables:

`GIT_TRACE=1`
`GIT_TRACE_CURL=1`
`GCM_TRACE=1`

Note that the output from `GIT_TRACE_CURL` can be quite verbose. It should not contain any secrets; however, it's recommended to double check before uploading out of an abundance of caution.

**Comment by mkolakow** [comment]
Thank you so much for the response!! Yes, indeed, our environment is relatively locked down (which is a good thing). We do basically have single sign on for most things. For example, once logged into to Windows, I can access Sql Server Management Studio w/o entering a user/pass. It just works. 

I will enable those env variables and start capturing logs and update here with any findings. 

What I don't understand is why it works some of the time for me. Case in point, I signed on before getting to this message and did a quick `git pull` and it still works today. Haven't changed anything. However, as previously stated, I had it working last week and then it suddenly stopped. 

When all is done and we have this figured out, I should probably write a nice blog post or SO question on it! 

Thanks,
Michael

**Comment by mkolakow** [comment]
Hi - we are collecting more logs, but are also off tomorrow, so I probably won't have any new information until next week. There are some image encoding in the logs - can I delete those? I'll also sanitize project/url/etc names before uploading.

Thanks,
Michael

**Comment by ldennington** [comment]
Next week is fine. And it also should be fine to remove the image encoding.

**Comment by mkolakow** [comment]
Hi,

I added the following to my {{bash_profile}} so that it wouldn't print out to the console every time:
```
export GIT_TRACE=~/git.log
export GIT_TRACE_CURL=~/git-curl-trace.log
export GCM_TRACE=~/gcm-trace.log
```
In any case, I am attaching the two curl trace logs. The first is from Thursday, when it was still working. The second log is from today and sure enough is a non-working example. I signed in the exact same way I do every day and proceeded to do a git pull, at which point I got the "Auth failed" message. I've compared the two curl logs and don't see much of a difference. Do you?

**Thursday log when it worked**
[git-curl-trace.log](https://github.com/git-ecosystem/git-credential-manager/files/13070961/git-curl-trace.log)

**Monday log when it did not work**
[git-curl-trace-not-working.log](https://github.com/git-ecosystem/git-credential-manager/files/13070968/git-curl-trace-not-working.log)

Note: I did change a few urls, names, and some text within the image data just to be safe. Otherwise, it's pretty much left alone. 

I also have the other two logs, but didn't see anything notable in them, besides which, you already have the two GCM trace logs above. The only thing that jumps out at me there is the 'store' vs 'erase' commands. I can attach them if you like, however. 

Thanks,
Michael

**Comment by mkolakow** [comment]
And go figure, I signed in today, no changes, and now magically it works! I can attach logs, but I'm guessing there's little point in doing so. 

One thing I considered: After logging in to laptop + vpn, I still have to log into internal websites through Chrome one time (first thing) in the morning. For example, if I go to our Confluence wiki, I won't be logged in. Then I'll click log in and have to enter my pin (same pin for my laptop and vpn). But that's one time only. Now, I will get timed out during the day, but all I have to do for the rest of the day is click log in and it automatically logs me in (no pin entry required). So:

* I don't know how SSO is set up here. Maybe there's something wonky there?
* I don't know what Windows is doing behind the scenes with those credentials (caching??)
* Perhaps I hadn't logged in once through chrome on the days git didn't work, but
  * I would expect a subsequent login to, say confluence, would fix that (and it doesn't).

Thanks,
Michael

**Comment by mjcheetham** [comment]
Looking at the differences between the working and not-working curl logs, I see this in the latter:

```logs
..
07:43:35.359272 http.c:845              == Info: NTLM handshake rejected
07:43:35.359272 http.c:845              == Info: Authentication problem. Ignoring this.
..
```

Looks like there is indeed something weird going on with NTLM between the client and server. GCM does not actually perform NTLM or Kerberos authentication itself, and just lets Git/curl handle it.

> One thing I considered: After logging in to laptop + vpn, I still have to log into internal websites through Chrome one time (first thing) in the morning. For example, if I go to our Confluence wiki, I won't be logged in.

This sounds like there's definitely an issue with your SSO. Is this affecting others in your environment or just you?

It may be possible that although you've managed to log-in to Windows, your machine didn't 'see' the domain controller so is using cached creds until the VPN connects, and then you _next_ perform a NTLM handshake? I'm not an NTLM expert, so this is just a guess.

Does this change if you are physically on the corporate network rather than via 
the VPN?

**Comment by mkolakow** [comment]
Good catch - I missed that when I was pouring through the logs. I also see the following before the handshake rejection:

working: `Recv header: HTTP/1.1 200 OK`
not-working: `Recv header: HTTP/1.1 401 Unauthorized`

To answer your first question, this affects everyone. The whole dev team has to get a temp one-day password every day for git, which is why I'm pushing hard to solve this. 

I suppose the next step is to figure out who even is in charge of SSO. I'm new, so that's been difficult. We're big (over 10,000 employees), but I've had a hard time getting straight answers on that front. 🤣  And for what it's worth, people on the team said it "used to work" like a year ago. And because we can all log on to the ADO server with no problem, I pursued a solution from our end. 

To answer your second question - no idea. I'm fully remote two states away. Most of the team is fully remote too. However, if I can get someone to do that testing I will. Even so, it still needs to be solved for all the fully remote folks.

Thanks,
Michael

**Comment by ldennington** [comment]
> And for what it's worth, people on the team said it "used to work" like a year ago.

Something that may be worth trying is installing an older version of Git/GCM to see if it changes anything. If you still run into the same problems, it indicates that your network environment has likely changed in some way that caused this issue to manifest. Here's a link to a version of Git from about a year and a half ago you can try:

https://github.com/git-for-windows/git/releases/tag/v2.37.0.windows.1

Please let us know what you find.

Also, would it be possible to ask the maintainer of your Azure DevOps Server instance to see if they can see anything in the logs (specifically the web server logs, which handle NTLM)? The message from CURL that the handshake was rejected is all we can see on the client side - the server is rejecting your connection for some reason.
